lib/mauth/request_and_response.rb

```diff
@@ -23,6 +39,67 @@ def string_to_sign(more_attributes)
       Digest::SHA512.hexdigest(string)
     end
 
+    # the string to sign for V2 protocol will be (where LF is line feed character)
+    # for requests:
+    #   string_to_sign =
+    #     http_verb + <LF> +
+    #     resource_url_path (no host, port or query string; first "/" is included) + <LF> +
+    #     request_body_digest + <LF> +
+    #     app_uuid + <LF> +
+    #     current_seconds_since_epoch + <LF> +
+    #     encoded_query_params
+    #
+    # for responses:
+    #   string_to_sign =
+    #     status_code_string + <LF> +
+    #     response_body_digest + <LF> +
+    #     app_uuid + <LF> +
+    #     current_seconds_since_epoch
+    def string_to_sign_v2(more_attributes)
+      attributes_for_signing = self.attributes_for_signing.merge(more_attributes)
+
+      # lazy instantiation of body digest to avoid hashing request bodies
+      # three times because we call string to sign three times.
+      if attributes_for_signing[:body]
+        attributes_for_signing[:body_digest] ||= Digest::SHA512.hexdigest(attributes_for_signing[:body].to_s)
+      end
+      attributes_for_signing[:encoded_query_params] = encode_query_string(attributes_for_signing[:query_string].to_s)
+
+      missing_attributes = self.class::SIGNATURE_COMPONENTS_V2.reject do |key|
+        attributes_for_signing.dig(key)
+      end
+
+      missing_attributes.delete(:body_digest) # body may be omitted
+      if missing_attributes.any?
```

Comment:
I'm wondering if there should be specs testing all permutations of missing attributes for the string to sign (for v1 and v2).  Thoughts?

Reply:
I think it makes sense. I'm writing specs for the `Signable` module to address this (and https://github.com/mdsol/mauth-client-ruby/pull/25#discussion_r293524108)

Reply:
I'd recommend the use of mutant https://github.com/mbj/mutant
It will force your specs to really cover all paths. We have used it for other foundational gems   where we do not want to take any chances.

Reply:
Feel free to ping me on my email if you wanted help with mutant use.